File: apps/web/src/features/investments/utils/brokerage-submission.ts
import { normalizeDecimal } from "@/lib/financial-calculations/decimal"
import type { SubmissionAttempt } from "@/features/accounts/utils/refund-submission"
import { resolveSubmissionAttempt } from "@/features/accounts/utils/refund-submission"

const decimal = (value: string | null | undefined, zero = false) =>
  normalizeDecimal(value?.trim() || (zero ? "0" : ""))

export function brokerageTradeFingerprint(input: {
  side: "buy" | "sell"; accountId: string; assetId: string; quantity: string;
  unitPrice: string; occurredAt: string; notes: string; fees: string; accountFxRate: string | null
}) {
  return JSON.stringify([input.side,input.accountId,input.assetId,decimal(input.quantity),decimal(input.unitPrice),input.occurredAt,input.notes.trim()||null,decimal(input.fees,true),decimal(input.accountFxRate)])
}

export function brokerageDividendFingerprint(input: {
  mode: "cash" | "full" | "partial"; accountId: string; assetId: string; gross: string;
  tax: string; fees: string; occurredAt: string; notes: string; unitPrice?: string; reinvestedAmount?: string
}) {
  return JSON.stringify([input.mode,input.accountId,input.assetId,decimal(input.gross),decimal(input.tax,true),decimal(input.fees,true),input.occurredAt,input.notes.trim()||null,input.mode === "cash" ? null : decimal(input.unitPrice),input.mode === "partial" ? decimal(input.reinvestedAmount) : null])
}

export const brokerageAttempt = (current: SubmissionAttempt | null, fingerprint: string) =>
  resolveSubmissionAttempt(current, fingerprint)

File: apps/mobile/lib/accounts/brokerage/brokerage_submission.dart
import 'dart:convert';
import '../../core/decimals.dart';
import '../../core/local_datetime.dart';
import 'brokerage_models.dart';
import 'brokerage_valuation.dart';

String brokerageTradeFingerprint(String accountId, TradeFormValues v) =>
    jsonEncode([
      v.side.name,
      accountId,
      v.assetId,
      D.normalize(v.quantity),
      D.normalize(v.unitPrice),
      localDateTimeInputToIso(v.occurredAt),
      v.notes.trim().isEmpty ? null : v.notes.trim(),
      D.normalize(v.fees.trim().isEmpty ? '0' : v.fees),
      D.normalize(v.accountFxRate),
    ]);

String brokerageDividendFingerprint({
  required String accountId,
  required String assetId,
  required DividendMode mode,
  required String gross,
  required String tax,
  required String fees,
  required String occurredAt,
  String? notes,
  String? unitPrice,
  String? reinvestedAmount,
}) => jsonEncode([
  mode.name,
  accountId,
  assetId,
  D.normalize(gross),
  D.normalize(tax.trim().isEmpty ? '0' : tax),
  D.normalize(fees.trim().isEmpty ? '0' : fees),
  localDateTimeInputToIso(occurredAt),
  notes?.trim().isEmpty == false ? notes!.trim() : null,
  mode == DividendMode.cash ? null : D.normalize(unitPrice),
  mode == DividendMode.partial ? D.normalize(reinvestedAmount) : null,
]);
